compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -875,6 +864,14 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                 if let &Const::Yes(span) = constness {
                     self.err_handler().emit_err(error(span, "`const`", true));
                 }
+
+                self.visit_vis(&item.vis);
+                self.visit_ident(item.ident);
+                self.with_tilde_const(None, |this| this.visit_generics(generics));
```

Comment:
Currently permitting `~const` trait bounds on inherent impls not to regress [rfc-2632-const-trait-impl/tilde_const_on_impl_bound.rs](https://github.com/rust-lang/rust/blob/master/tests/ui/rfcs/rfc-2632-const-trait-impl/tilde_const_on_impl_bound.rs). Not sure if the following code should at some point compile:

```rs
struct Struct;
#[const_trait] trait Trait {}
impl<T: ~const Trait> Struct {
    const fn f(_: T) {}
}
```

If I were to straight up deny `~const` here, it'd obviously fix #117004. I haven't tested yet if the `~const` lowering code can properly lower the code above.

Reply:
We should probably deny this first. We don't really *need* this to work and making it work certainly adds to the overhead. Perhaps a followup PR? Or I can do that. Thanks for working on this! :)

Reply:
I think we should deny `~const` here. If we want to be able to use `Trait` in `const fn f`, I think the user should need to write:

```
struct Struct;
#[const_trait] trait Trait {}
impl<T: Trait> Struct {
    const fn f(_: T) where T: ~const Trait {}
}
```

at least in my opinion.